OS X - Error compiling temporal_convolution_tbc

OS: OS X 10.13 High Sierra
CUDA: v. 9.0


When trying to compile fairseq-py on OSX, I'm getting the error:
> error: non-const lvalue reference to type 'at::Tensor' cannot bind to a temporary of type 'int'

I've tried using g++ , with similar results:
>error: invalid initialization of non-const reference of type 'at::Tensor&' from an rvalue of type 'int'


Here's the full output:

`
$ CC=clang++ python setup.py build
running build
running build_py
generating /var/folders/1f/z12pg_yx2194lgdvlb_kprb40000gp/T/tmpemf541ha/_temporal_convolution_tbc.c
running build_ext
building '_temporal_convolution_tbc' extension
creating Users
creating Users/mastover
creating Users/mastover/git
creating Users/mastover/git/fairseq-py
creating Users/mastover/git/fairseq-py/fairseq
creating Users/mastover/git/fairseq-py/fairseq/clib
creating Users/mastover/git/fairseq-py/fairseq/clib/temporal_convolution_tbc
clang++ -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -fwrapv -O2 -Wall -Wstrict-prototypes -march=core2 -mtune=haswell -mssse3 -ftree-vectorize -fPIC -fPIE -fstack-protector-strong -O2 -pipe -march=core2 -mtune=haswell -mssse3 -ftree-vectorize -fPIC -fPIE -fstack-protector-strong -O2 -pipe -DWITH_CUDA -I/Users/mastover/anaconda3/lib/python3.6/site-packages/torch/utils/ffi/../../lib/include -I/Users/mastover/anaconda3/lib/python3.6/site-packages/torch/utils/ffi/../../lib/include/TH -I/Users/mastover/anaconda3/lib/python3.6/site-packages/torch/utils/ffi/../../lib/include/THC -I/usr/local/cuda/include -I/Developer/NVIDIA/CUDA-9.0/include -I/Developer/NVIDIA/CUDA-8.0/include -I/Users/mastover/anaconda3/include/python3.6m -c _temporal_convolution_tbc.c -o ./_temporal_convolution_tbc.o -std=c++11
clang: warning: treating 'c' input as 'c++' when in C++ mode, this behavior is deprecated [-Wdeprecated]
clang++ -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -fwrapv -O2 -Wall -Wstrict-prototypes -march=core2 -mtune=haswell -mssse3 -ftree-vectorize -fPIC -fPIE -fstack-protector-strong -O2 -pipe -march=core2 -mtune=haswell -mssse3 -ftree-vectorize -fPIC -fPIE -fstack-protector-strong -O2 -pipe -DWITH_CUDA -I/Users/mastover/anaconda3/lib/python3.6/site-packages/torch/utils/ffi/../../lib/include -I/Users/mastover/anaconda3/lib/python3.6/site-packages/torch/utils/ffi/../../lib/include/TH -I/Users/mastover/anaconda3/lib/python3.6/site-packages/torch/utils/ffi/../../lib/include/THC -I/usr/local/cuda/include -I/Developer/NVIDIA/CUDA-9.0/include -I/Developer/NVIDIA/CUDA-8.0/include -I/Users/mastover/anaconda3/include/python3.6m -c /Users/mastover/git/fairseq-py/fairseq/clib/temporal_convolution_tbc/temporal_convolution_tbc.cpp -o ./Users/mastover/git/fairseq-py/fairseq/clib/temporal_convolution_tbc/temporal_convolution_tbc.o -std=c++11
/Users/mastover/git/fairseq-py/fairseq/clib/temporal_convolution_tbc/temporal_convolution_tbc.cpp:70:21: error: non-const lvalue reference to type 'at::Tensor' cannot bind to a temporary of type 'int'
      at::addmm_out(1, O, 1, I, W, O);
                    ^
/Users/mastover/anaconda3/lib/python3.6/site-packages/torch/utils/ffi/../../lib/include/ATen/Functions.h:1203:43: note: passing argument to parameter 'result' here
static inline Tensor & addmm_out(Tensor & result, const Tensor & self, const Tensor & mat1, const Tensor & mat2, Scalar beta, Scalar alpha) {
                                          ^
/Users/mastover/git/fairseq-py/fairseq/clib/temporal_convolution_tbc/temporal_convolution_tbc.cpp:111:21: error: non-const lvalue reference to type 'at::Tensor' cannot bind to a temporary of type 'int'
      at::addmm_out(1, dI, 1, dO, weight[k].t(), dI);
                    ^
/Users/mastover/anaconda3/lib/python3.6/site-packages/torch/utils/ffi/../../lib/include/ATen/Functions.h:1203:43: note: passing argument to parameter 'result' here
static inline Tensor & addmm_out(Tensor & result, const Tensor & self, const Tensor & mat1, const Tensor & mat2, Scalar beta, Scalar alpha) {
                                          ^
/Users/mastover/git/fairseq-py/fairseq/clib/temporal_convolution_tbc/temporal_convolution_tbc.cpp:124:21: error: non-const lvalue reference to type 'at::Tensor' cannot bind to a temporary of type 'int'
      at::addmm_out(1, dW, 1, I, dO, dW);
                    ^
/Users/mastover/anaconda3/lib/python3.6/site-packages/torch/utils/ffi/../../lib/include/ATen/Functions.h:1203:43: note: passing argument to parameter 'result' here
static inline Tensor & addmm_out(Tensor & result, const Tensor & self, const Tensor & mat1, const Tensor & mat2, Scalar beta, Scalar alpha) {
                                          ^
/Users/mastover/git/fairseq-py/fairseq/clib/temporal_convolution_tbc/temporal_convolution_tbc.cpp:129:20: error: reference to type 'const at::Tensor' could not bind to an rvalue of type 'int'
  at::sum_out(tmp, 0, dBias);
                   ^
/Users/mastover/anaconda3/lib/python3.6/site-packages/torch/utils/ffi/../../lib/include/ATen/Functions.h:1044:64: note: passing argument to parameter 'self' here
static inline Tensor & sum_out(Tensor & result, const Tensor & self, int64_t dim, bool keepdim) {
                                                               ^
4 errors generated.
Traceback (most recent call last):
  File "/Users/mastover/anaconda3/lib/python3.6/distutils/unixccompiler.py", line 118, in _compile
    extra_postargs)
  File "/Users/mastover/anaconda3/lib/python3.6/distutils/ccompiler.py", line 909, in spawn
    spawn(cmd, dry_run=self.dry_run)
  File "/Users/mastover/anaconda3/lib/python3.6/distutils/spawn.py", line 36, in spawn
    _spawn_posix(cmd, search_path, dry_run=dry_run)
  File "/Users/mastover/anaconda3/lib/python3.6/distutils/spawn.py", line 159, in _spawn_posix
    % (cmd, exit_status))
distutils.errors.DistutilsExecError: command 'clang++' failed with exit status 1

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/mastover/anaconda3/lib/python3.6/site-packages/cffi/ffiplatform.py", line 49, in _build
    dist.run_command('build_ext')
  File "/Users/mastover/anaconda3/lib/python3.6/distutils/dist.py", line 974, in run_command
    cmd_obj.run()
  File "/Users/mastover/anaconda3/lib/python3.6/site-packages/setuptools/command/build_ext.py", line 75, in run
    _build_ext.run(self)
  File "/Users/mastover/anaconda3/lib/python3.6/site-packages/Cython/Distutils/old_build_ext.py", line 185, in run
    _build_ext.build_ext.run(self)
  File "/Users/mastover/anaconda3/lib/python3.6/distutils/command/build_ext.py", line 339, in run
    self.build_extensions()
  File "/Users/mastover/anaconda3/lib/python3.6/site-packages/Cython/Distutils/old_build_ext.py", line 193, in build_extensions
    self.build_extension(ext)
  File "/Users/mastover/anaconda3/lib/python3.6/site-packages/setuptools/command/build_ext.py", line 196, in build_extension
    _build_ext.build_extension(self, ext)
  File "/Users/mastover/anaconda3/lib/python3.6/distutils/command/build_ext.py", line 533, in build_extension
    depends=ext.depends)
  File "/Users/mastover/anaconda3/lib/python3.6/distutils/ccompiler.py", line 574, in compile
    self._compile(obj, src, ext, cc_args, extra_postargs, pp_opts)
  File "/Users/mastover/anaconda3/lib/python3.6/distutils/unixccompiler.py", line 120, in _compile
    raise CompileError(msg)
distutils.errors.CompileError: command 'clang++' failed with exit status 1

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "setup.py", line 69, in <module>
    'build_py': build_py_hook,
  File "/Users/mastover/anaconda3/lib/python3.6/distutils/core.py", line 148, in setup
    dist.run_commands()
  File "/Users/mastover/anaconda3/lib/python3.6/distutils/dist.py", line 955, in run_commands
    self.run_command(cmd)
  File "/Users/mastover/anaconda3/lib/python3.6/distutils/dist.py", line 974, in run_command
    cmd_obj.run()
  File "/Users/mastover/anaconda3/lib/python3.6/distutils/command/build.py", line 135, in run
    self.run_command(cmd_name)
  File "/Users/mastover/anaconda3/lib/python3.6/distutils/cmd.py", line 313, in run_command
    self.distribution.run_command(command)
  File "/Users/mastover/anaconda3/lib/python3.6/distutils/dist.py", line 974, in run_command
    cmd_obj.run()
  File "setup.py", line 49, in run
    conv_tbc.build()
  File "/Users/mastover/anaconda3/lib/python3.6/site-packages/torch/utils/ffi/__init__.py", line 164, in build
    _build_extension(ffi, cffi_wrapper_name, target_dir, verbose)
  File "/Users/mastover/anaconda3/lib/python3.6/site-packages/torch/utils/ffi/__init__.py", line 100, in _build_extension
    ffi.compile(tmpdir=tmpdir, verbose=verbose, target=libname)
  File "/Users/mastover/anaconda3/lib/python3.6/site-packages/cffi/api.py", line 684, in compile
    compiler_verbose=verbose, debug=debug, **kwds)
  File "/Users/mastover/anaconda3/lib/python3.6/site-packages/cffi/recompiler.py", line 1484, in recompile
    compiler_verbose, debug)
  File "/Users/mastover/anaconda3/lib/python3.6/site-packages/cffi/ffiplatform.py", line 20, in compile
    outputfilename = _build(tmpdir, ext, compiler_verbose, debug)
  File "/Users/mastover/anaconda3/lib/python3.6/site-packages/cffi/ffiplatform.py", line 56, in _build
    raise VerificationError('%s: %s' % (e.__class__.__name__, e))
cffi.error.VerificationError: CompileError: command 'clang++' failed with exit status 1
`

--- a/fairseq/clib/temporal_convolution_tbc/temporal_convolution_tbc.cpp
+++ b/fairseq/clib/temporal_convolution_tbc/temporal_convolution_tbc.cpp
@@ -67,7 +67,7 @@ extern "C" void TemporalConvolutionTBC_forward(
       auto W = weight[k];
       auto I = input.narrow(0, iShift, t).view({t * batchSize, inputPlanes});
       auto O = output.narrow(0, oShift, t).view({t * batchSize, outputPlanes});
-      at::addmm_out(1, O, 1, I, W, O);
+      O.addmm_(I, W);
     }
   }
 }
@@ -108,7 +108,7 @@ extern "C" void TemporalConvolutionTBC_backward(
     if (t > 0) {
       auto dO = dOutput.narrow(0, oShift, t).view({t * batchSize, outputPlanes});
       auto dI = dInput.narrow(0, iShift, t).view({t * batchSize, inputPlanes});
-      at::addmm_out(1, dI, 1, dO, weight[k].t(), dI);
+      dI.addmm_(dO, weight[k].t());
     }
   }
 
@@ -121,10 +121,10 @@ extern "C" void TemporalConvolutionTBC_backward(
       auto dW = dWeight[k];
       auto dO = dOutput.narrow(0, oShift, t).view({t * batchSize, outputPlanes});
       auto I = input.narrow(0, iShift, t).view({t * batchSize, inputPlanes}).t();
-      at::addmm_out(1, dW, 1, I, dO, dW);
+      dW.addmm_(I, dO);
     }
   }
 
   auto tmp = dOutput.sum(0, false);
-  at::sum_out(tmp, 0, dBias);
+  dBias.assign_(tmp.sum(0));
 }

--- a/tests/test_convtbc.py
+++ b/tests/test_convtbc.py
@@ -1,0 +1,51 @@
+# Copyright (c) 2017-present, Facebook, Inc.
+# All rights reserved.
+#
+# This source code is licensed under the license found in the LICENSE file in
+# the root directory of this source tree. An additional grant of patent rights
+# can be found in the PATENTS file in the same directory.
+#
+
+import torch
+import unittest
+from fairseq.modules import ConvTBC
+import torch.nn as nn
+from torch.autograd import Variable, gradcheck
+
+
+class TestConvTBC(unittest.TestCase):
+
+    def test_convtbc(self):
+        # ksz, in_channels, out_channels
+        conv_tbc = ConvTBC(4, 5, kernel_size=3, padding=1)
+        # out_channels, in_channels, ksz
+        conv1d = nn.Conv1d(4, 5, kernel_size=3, padding=1)
+
+        conv_tbc.weight.data.copy_(conv1d.weight.data.transpose(0, 2))
+        conv_tbc.bias.data.copy_(conv1d.bias.data)
+
+        input_tbc = Variable(torch.randn(7, 2, 4), requires_grad=True)
+        input1d = Variable(input_tbc.data.transpose(0, 1).transpose(1, 2), requires_grad=True)
+
+        output_tbc = conv_tbc(input_tbc)
+        output1d = conv1d(input1d)
+
+        self.assertAlmostEqual(output_tbc.data.transpose(0, 1).transpose(1, 2), output1d.data)
+        
+        grad_tbc = torch.randn(output_tbc.size())
+        grad1d = grad_tbc.transpose(0, 1).transpose(1, 2).contiguous()
+
+        output_tbc.backward(grad_tbc)
+        output1d.backward(grad1d)
+
+        self.assertAlmostEqual(conv_tbc.weight.grad.data.transpose(0, 2), conv1d.weight.grad.data)
+        self.assertAlmostEqual(conv_tbc.bias.grad.data, conv1d.bias.grad.data)
+        self.assertAlmostEqual(input_tbc.grad.data.transpose(0, 1).transpose(1, 2), input1d.grad.data)
+
+    def assertAlmostEqual(self, t1, t2):
+        self.assertEqual(t1.size(), t2.size(), "size mismatch")
+        self.assertLess((t1 - t2).abs().max(), 1e-4)
+
+
+if __name__ == '__main__':
+    unittest.main()

